ecs-agent/wsclient/client.go

```diff
@@ -565,3 +589,42 @@ func permissibleCloseCode(err error) bool {
 		websocket.CloseGoingAway,         // websocket error code 1001
 		websocket.CloseInternalServerErr) // websocket error code 1011
 }
+
+// newDisconnectTimeoutHandler returns new timer object to disconnect from server connection start time, with goroutine
+// to disconnect from client.
+func (cs *ClientServerImpl) newDisconnectTimeoutHandler(startTime time.Time,
+	metricName string,
+	disconnectTimeout time.Duration,
+	disconnectJitterMax time.Duration) *time.Timer {
+
+	maxConnectionDuration := retry.AddJitter(disconnectTimeout, disconnectJitterMax)
+	logger.Info(("Websocket connection established."), logger.Fields{
+		"URL":                    cs.URL,
+		"ConnectTime":            startTime.Format(dateTimeFormat),
+		"ExpectedDisconnectTime": startTime.Add(disconnectTimeout).Format(dateTimeFormat),
+	})
+
+	timer := time.AfterFunc(maxConnectionDuration, func() {
+		err := cs.CloseClient(startTime, maxConnectionDuration)
+		cs.MetricsFactory.New(metricName).Done(err)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Attempted disconnecting; client already closed. %s", err))
+		}
+	})
+	return timer
+}
+
+// closeClient will attempt to close the provided client, retries are not recommended
+// as failure modes for this are when client is not found or already closed.
+func (cs *ClientServerImpl) CloseClient(startTime time.Time, timeoutDuration time.Duration) error {
+	logger.Warn(("Closing connection"), logger.Fields{
+		"ConnectionStartTime":  startTime.Format(dateTimeFormat),
+		"MaxDisconnectionTime": startTime.Add(timeoutDuration).Format(dateTimeFormat),
+	})
+	err := cs.WriteCloseMessage()
+	if err != nil {
```

Comment:
When i looked through the code for writecontrol, there seems to be x number of possible errors in the workflow. It is impossible to capture all of them. I would say, since we are not doing any error check here, but just logging it and returning the error, we should probably remove the log statement "client already closed" and let the error say for itself.

Reply:
Sure sounds good! If possible, let's use logger.Fields here as well.